### ISSUE TITLE
Content Security Policy meta tag breaks Admin Interface if 'unsafe-eval' is not allowed

### ISSUE THREAD

**Issue by tackerm** [sulu/sulu#4387]
| Q | A
| --- | ---
| Bug? | yes
| New Feature? | no
| Sulu Version | 1.6.24
| Browser Version | Firefox Quantum 64.0

#### Actual Behavior

Adding a Content Security Policy meta tag in the frontend breaks the Sulu Admin interface if `script-src 'unsafe-eval';` is not enabled.
Sulu Admin works fine until a preview of the frontend (with the CSP meta tag) is being displayed for the first time - after that, the CSP prevents execution of some code that is required by the Admin interface.

#### Expected Behavior

Sulu Admin should not require `script-src 'unsafe-eval';`, as this is a potential security issue.

#### Steps to Reproduce

1. Add the following meta tag to the frontend template: `
<meta http-equiv="Content-Security-Policy" content="default-src 'self'; font-src 'self' ; img-src 'self'; object-src 'self'; script-src 'self'; style-src 'self'; frame-src 'self';" />`

1. Log in to Sulu Admin, open the editor that loads a preview of the frontend. The editor fails to load.

1. After changing the CSP tag to  `
<meta http-equiv="Content-Security-Policy" content="default-src 'self'; font-src 'self' ; img-src 'self'; object-src 'self'; script-src 'self' 'unsafe-eval'; style-src 'self'; frame-src 'self';" />`, the editor works again.

#### Possible Solutions

Firefox console prints the following:
```
aura_1: call to Function() blocked by CSP husky.js:28823:5
	Logger.prototype.write http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:28823:5
	Logger.prototype.error http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:28835:5
	Component.load/< http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:16130:11
	execCb http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:1636:24
	check http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:870:43
	enable http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:1141:17
	init http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:778:21
	localRequire/< http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:1411:25

aura_1: Error loading component: preview EvalError: "call to Function() blocked by CSP"
	template http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:13053:16
	templates http://localhost:4572/bundles/suluadmin/js/aura_extensions/default-extension.js?v=develop:44:38
	collect http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:11946:20
	forEach http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:11934:13
	collect http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:11945:5
	<anonymous> http://localhost:4572/bundles/suluadmin/js/aura_extensions/default-extension.js?v=develop:43:21
	invokeCallbacks http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:15892:21
	each http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:2714:14
	invokeCallbacks http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:15890:7
	invokeWithCallbacks http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:15913:16
	invokeWithCallbacks http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:16017:14
	Component http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:15967:7
	child http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:16030:29
	child http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:16030:29
	load http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:16118:30
	execCb http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:1636:24
	check http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:870:43
	enable http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:1141:17
	init http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:778:21
	localRequire http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:1411:25
 husky.js:28823:5
	Logger.prototype.write http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:28823:5
	Logger.prototype.error http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:28835:5
	Component.load/< http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:16067:9
	fire http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:5105:10
	fireWith http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:5217:7
	Deferred/</deferred[tuple[0]] http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:5306:34
	Component.load/< http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:16131:11
	execCb http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:1636:24
	check http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:870:43
	enable http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:1141:17
	init http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:778:21
	localRequire/< http://localhost:4572/bundles/suluadmin/js/vendor/husky/husky.js:1411:25
```

The problem seems to be related to the Husky library. I am not familiar with Husky, but it looks like a dev-dependency to me..? If that is the case, would it be safe to remove it?

**Comment by alexander-schranz** [comment]
@tackerm Thank you for your issue! [Husky](http://github.com/massiveart/husky) is our UI library sulu 1.x is based on so we can't remove it. If I understand the diff correctly is that you added `unsafe-eval` to get it work again:

The line of the stacktrace seems to point to the following file: https://github.com/massiveart/husky/blob/0.30.5/bower_components/aura/lib/logger.js#L43 of the aura.js logger and https://github.com/massiveart/husky/blob/0.30.5/bower_components/underscore/underscore.js#L1206 of the underscore library
and  not sure what is false with them.

A workaround for you could currently be add a check if the page is rendered in preview or not by using:

```twig
{% if not app.request.attributes.get('preview') %}
     <meta http-equiv="Content-Security-Policy" ...>
{% endif %}
```

**Comment by tackerm** [comment]
Ah, I confused it with [another](https://github.com/typicode/husky) husky library, that's why I suggested to remove it :)

Looking at the lines you mentioned, the `new Function(...)` call in the underscore library (https://github.com/massiveart/husky/blob/5b4cceb14179fa26c347e371530cd05b8c77026f/bower_components/underscore/underscore.js#L1206) seems to be the source of the problem - this is (like `eval()`) considered unsafe and blocked by default CSP configurations. (See also the explanations here: https://developers.google.com/web/fundamentals/security/csp/#eval_too)

So, the proposed solution would be to re-write that part of the code.

Regarding the workaround, I am not sure how viable this is. From what I understand, the `preview` attribute is set via the request header? If so, an attacker could probably disable CSP simply by forging a request that includes the `preview` attribute - which would then allow them to execute malicious code.

**Comment by tackerm** [comment]
The latest underscore version (1.9.1) still uses the `new Function()` call, and it is unlikely that this is going to change: https://github.com/jashkenas/underscore/issues/906

I think there are two separate issues here:

1. Relying on `new Function()` is generally considered unsafe. It looks like precompiling the templates is the way to go around this - this is also suggested by the [Google Dev article](https://developers.google.com/web/fundamentals/security/csp/#eval_too) linked above.

1. The admin section should be "immune" against whatever is in the frontend template. If a meta tag in the frontend can break the admin area, the admin area is probably also vulnerable to other XSS attacks via the preview feature (e.g. if there is malicious code in the frontend for whatever reason). This could maybe be adressed by [sandboxing](https://www.html5rocks.com/en/tutorials/security/sandboxed-iframes/) the preview. (In addition, the admin area could maybe also define it's own CSP record for additional security...)

**Comment by alexander-schranz** [comment]
@tackerm 

1. I'm not sure how this can be implemented. If you are familiar with it we are really open for pull requests. 

2. The preview is rendered in a seperate iframe so more seperation is not really possible in case of html between admin and preview. I think the problem here as the browsers sees one page set the CSP record it is used for all request on that domain, which make in case of security totally sense. Not sure what happens if both define CSP you could try it in: https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/AdminBundle/Resources/views/Admin/index.html.twig#L8

**Comment by tackerm** [comment]
1. I'm afraid I won't be much help here - I'm not really a frontend guy. I'm happy to help with ideas and suggestions though.

1. I quickly tried setting a CSP in the admin, it didn't solve the problem. Quite the opposite actually, it is propagated to the contents of the iframe and basically breaks the preview. This propagation from parent -> embedded is what I would expect, but I find it strange that the CSP of an embedded document is applied to the parent document. This made me think that it might actually be a browser issue - and voilà, I can not reproduce the problem with Google Chrome :man_facepalming: 

So, bottom line: The templating / `new Function()` topic is still something that should be looked into, but maybe for a different reason than originally described. I'll leave it up to you @alexander-schranz whether you want to keep this issue open or not - if you prefer to close it, that's fine for me.

**Comment by alexander-schranz** [comment]
This seems only be an issue in the old `1.6` and seems not longer exist in `2.x` version. In 1.6 the issue leading is in the `underscore.js` version used by husky as linked above this can currently not be fixed here. Feedback here is still welcome! The good thing is that you still can set it on the website and just do something like to avoid setting the preview for the admin does fix it:

```twig
{% if app.request.attributes.get('preview') ==  false %}
     <meta ..>
{% endif %}
```

For Sulu 2 versions this is not longer needed there you can set the meta CSP without any issues.

**Comment by RealAestan** [comment]
There is an issue with the usage of the ajv library regarding the Content-Security-Policy header and the necessity to add 'unsafe-eval'. I think that there is a way to fix it :

- https://ajv.js.org/security.html#content-security-policy
- https://ajv.js.org/guide/environments.html#browsers
- https://github.com/ajv-validator/ajv/issues/1640

**Comment by alexander-schranz** [comment]
@RealAestan Thank you for the feedback. I'm currently not sure if the documented way work for dynamic external schemas. It seems the only workaround is compiling the schemas before and so ajv isn't used to validate the schemas. Or did you get a way to get this work?
Currently the ajv is created here: https://github.com/sulu/sulu/blob/2.x/src/Sulu/Bundle/AdminBundle/Resources/js/utils/Ajv/createAjv.js

**Comment by RealAestan** [comment]
@alexander-schranz I'm not sure. I never used this library so I don't know the implications, from what I see in the documentation you could use something like this
```
  formats: customFormats,
  code: {
    source: true,
    formats: _`require("./formats")`,
  }
```
to have a standalone code generation that does not use unsafe eval. Although, take it with a grain of salt since I didn't try this. Could you test this since you know the implications of this library better than I do ?
Though it's absolutely not urgent, since we can have a different Content-Security-Policy header for the admin and the front.

**Comment by alexander-schranz** [comment]
@RealAestan the require formats don't work in our case the formats do not exist as files which can be required, they are generated external by the api and returned. Thats why I think the current solutions provided by ajv for CSP do not work in our case. Sadly I'm not so deep in ajv, maybe @luca-rath know more about this if something form the docs would be possible. /cc @luca-rath
